[server] Allow to add and remove user tokens

jjmerchante opened the pull request:
This pull requests includes a new model named `Credential` that allows users to add tokens for every data source.
There new methods and GraphQL mutation to add, remove or delete data source credentials.

Each user  can only use its own tokens.

This PR also removes `apiToken` parameter from `fetchGitHubOwnerRepos` and select a user token (if exists) for that job.

Include also an Authorization header with the Bestiary JWT Token for each call from the UI.

Fixes https://github.com/chaoss/grimoirelab-bestiary/issues/94

jjmerchante committed:
[db] Create credentials model to add datasource tokens

This commits includes a new model named `Credential` that allows to
add tokens related to a user and a data source.
Include also methods to find credentials by id, add new credentials
or delete them.
Tokens are encrypted in database using django SECRET_KEY

Signed-off-by: JJMerchante <jj.merchante@gmail.com>

jjmerchante committed:
[api] Include methods to add and delete credentials

This commits includes new methods to manage credentials. It allows
adding new credential given a user, a datasource name and a token.
On the other side, it allows a user to remove one of its credentials
given a credential id.

Signed-off-by: JJMerchante <jj.merchante@gmail.com>

jjmerchante committed:
[schema] Include graphql methods to manage credentials

This commits allows a user to add a new credential given a
datasource and a token, delete one given its ID and list
its own credentials.
It also removes the apiToken from fetchGitHubOwnerRepos
and include an Authorization header with the Bestiary JWT Token.

Signed-off-by: JJMerchante <jj.merchante@gmail.com>

evamillan commented:
I've tested it from the frontend and the queries and mutations work, but I get different errors from the one on the `test_authentication` tests when I'm not logged in. In the `addCredentials` mutation, the error message is `Cannot assign \"<SimpleLazyObject: <django.contrib.auth.models.AnonymousUser object at 0x7f6d446842e0>>\": \"Credential.user\" must be a \"User\" instance.` and in the `credentials` query instead of an error it returns  `{
  "data": {
    "credentials": {
      "entities": null
    }
  }
}`

evamillan commented:
One more thing, I think the tokens are not encrypted on the database, they should be obfuscated in some way.

jjmerchante commented:
> I've tested it from the frontend and the queries and mutations work, but I get different errors from the one on the `test_authentication` tests when I'm not logged in. In the `addCredentials` mutation, the error message ...

Thank you, I didn't notice what happens when the GraphQL API is called with a non-authenticated user. When it request a list of repositories I changed it to an empty list, do you prefer an error? I am not sure what is the best UX scenario. For the error when a non-authenticated user adds a new credential I changed the error text.

> One more thing, I think the tokens are not encrypted on the database, they should be obfuscated in some way.

I need to think a bit about this part, maybe include a key in the settings to encrypt/decrypt the value of the token.

evamillan commented:
> When it request a list of repositories I changed it to an empty list, do you prefer an error? I am not sure what is the best UX scenario.

It can work either way on the UI since it will only be accessible to users that are logged in, but it could be helpful to return an error in case someone wants to use the GraphQL API without the interface.

jjmerchante commented:
I made the following changes:
- When a user is not authenticated, it now returns an error when requesting their tokens.
- Credentials now must have a `name`.
- Tokens are now encrypted in the database using the Django `SECRET_KEY`.

sduenas commented on bestiary/core/models.py:
```diff
@@ -68,6 +71,33 @@ def pre_save(self, model_instance, add):
+    """Field to store encrypted data using SECRET_KEY.
+
+    The max length for this field is 79 chars which
+    are 184 encrypted chars using Fernet encryption
```
I think we should change this to AES 256. Although Fermet looks like a good choice I'd rather use the other one because many goberment agencies require that method.

sduenas reviewed (commented):
What do you think about changing to other method?

sduenas commented on bestiary/core/models.py:
```diff
@@ -344,3 +374,37 @@ def save(self, *args, **kwargs):
+                      help_text='user allowed to user this token.')
+    name = CharField(max_length=MAX_SIZE_TITLE_FIELD,
+                     help_text='Name of the token.')
+    token = EncryptedField(max_length=MAX_SIZE_CHAR_INDEX,
```
I think this is wrong. This is only a restriction for indices but token isn't going to be part of an index. This can be set to 256 chars without problems, so the full key is stored.

sduenas reviewed (approved):
LGTM
